### PR TITLE
Implement swipe and click actions

### PR DESCRIPTION
**Pull request by dpgraham**
* Add Kotlin to build.gradle and start using `.kt` files. `.kt` and `.java` are interoperable, so no concern about using both simultaneously
* Implement [Click Action](https://developer.android.com/reference/androidx/test/espresso/action/GeneralClickAction) and [Swipe Action](https://developer.android.com/reference/androidx/test/espresso/action/GeneralSwipeAction)
* Swipe action adds to the already existing `mobile: swipe` mobile endpoint. Can now pass parameters `{swiper, startCoordinates, endCoordinates, ...}`, 
* Click action adds a new `mobile: clickAction` mobile endpoint which takes parameters `{element, tapper, coordinatesProvider, precisionDescriber, inputDevice, buttonState}`
* Added JUnit tests to test the parsers

### CODE REVIEW

File: espresso-server/app/build.gradle [hunk 1]
```diff
@@ -44,6 +44,8 @@ dependencies {
     testImplementation 'org.mockito:mockito-core:2.8.9'
     testImplementation 'org.nanohttpd:nanohttpd-webserver:2.3.1'
     testImplementation 'org.robolectric:robolectric:4.0.2'
+    testImplementation "org.jetbrains.kotlin:kotlin-test:$kotlin_version"
```
Comment (mykola-mokhnach) [inline espresso-server/app/build.gradle]: where kotlin version is set?

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/GsonParserHelpers.kt [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+import com.google.gson.JsonObject
+import com.google.gson.JsonParseException
+
+class GsonParserHelpers {
+
+    inline fun <reified T : Enum<T>> parseEnum(jsonObj: JsonObject, propName: String, helperMessage: String = ""): T? {
+        val property = jsonObj.get(propName)
+        if (property != null) {
+            val propValueAsString = jsonObj.get(propName).asString.toUpperCase()
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/GsonParserHelpers.kt]: this might throw an NPE if prop value is null

Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/GsonParserHelpers.kt]: property.asString ...

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/GsonParserHelpers.kt [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+            try {
+                return enumValueOf<T>(propValueAsString)
+            } catch (e: Exception) {
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/GsonParserHelpers.kt]: can the exception class be narrowed down?

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/http/Router.java [hunk 4]
```diff
@@ -289,22 +293,29 @@ public BaseResponse route(String uri, Method method, Map<String, String> files)
...
-        }
+            // Parse the parameters
+            AppiumParams appiumParams;
+            if (postJson == null) {
+                appiumParams = new AppiumParams();
+            } else {
+                logger.debug(String.format("Got raw post data: %s", abbreviate(postJson, 300)));
+                try {
+                    appiumParams = paramClass.cast((new Gson()).fromJson(postJson, paramClass));
+                } catch (Exception e) {
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/http/Router.java]: can the exception class be narrowed down?

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/http/Router.java [hunk 5]
```diff
@@ -289,22 +293,29 @@ public BaseResponse route(String uri, Method method, Map<String, String> files)
...
+                    // If failed to parse params, throw an invalid argument exception
+                    return new AppiumResponse<>(AppiumStatus.INVALID_ARGUMENT, Log.getStackTraceString(e));
+                }
+            }
+            appiumParams.initUriMapping(uriParams);
 
-        // Create the result
-        try {
+            // Validate the sessionId
+            if (appiumParams.getSessionId() != null && !appiumParams.getSessionId().equals(Session.getGlobalSession().getId())) {
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/http/Router.java]: shouldn't this be checked first?

File: lib/commands/general.js [hunk 6]
```diff
@@ -198,6 +200,17 @@ commands.flashElement = async function (opts = {}) {
   });
 };
 
+/**
+ * Perform a 'GeneralClickAction' (https://developer.android.com/reference/androidx/test/espresso/action/GeneralClickAction)
+ */
+commands.mobileClickAction = async function (opts = {}) {
+  const {element} = assertRequiredOptions(opts, ['element']);
+  const {tapper, coordinatesProvider, precisionDescriber, inputDevice, buttonState} = opts;
```
Comment (mykola-mokhnach) [inline lib/commands/general.js]: this all can be extracted in a single line

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/GsonParserHelpers.kt [hunk 7]
```diff
@@ -0,0 +1,22 @@
+package io.appium.espressoserver.lib.helpers
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/GsonParserHelpers.kt]: copyright header is missing

File: espresso-server/app/build.gradle (hunk 1 above)
Comment (dpgraham) [reply espresso-server/app/build.gradle]: Added `ext.kotlin_version = 1.3.11`

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/GsonParserHelpers.kt (hunk 3 above)
Comment (dpgraham) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/GsonParserHelpers.kt]: Yep. Changed it to catch `IllegalArgumentException`

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/MobileSwipeParams.kt [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+
+            // Validate that exactly one of direction or swiper is set
+            if (swipeActionParams.direction != null && swipeActionParams.swiper != null) {
+                throw JsonParseException("Cannot set both 'direction' and 'swiper' for swipe action. Must set only one");
+            } else if (swipeActionParams.direction == null && swipeActionParams.swiper == null) {
+                throw JsonParseException("Must set one of 'direction' or 'swiper'");
+            }
+
+            // Deserialize 'startCoordinates'
+            val startCoordinates = gsonParserHelpers.parseEnum<GeneralLocation>(
```
[the remark this answers is not in the document]
Comment (dpgraham) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/MobileSwipeParams.kt]: Good idea. I added a default parameter to `parseEnum` that specifies the default value.

**Comment by dpgraham** [comment]
Fixes pushed

File: espresso-server/app/src/test/java/io/appium/espressoserver/test/model/MobileSwipeTest.kt [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+        val mobileSwipeActionParams = MobileSwipeActionParamsDeserializer()
+                .deserialize(jsonElement, null, null)
+        assertEquals(mobileSwipeActionParams.swiper, FAST);
+        assertEquals(mobileSwipeActionParams.startCoordinates, TOP_LEFT);
+        assertEquals(mobileSwipeActionParams.endCoordinates, TOP_RIGHT);
+        assertEquals(mobileSwipeActionParams.precisionDescriber, PINPOINT);
+    }
+
+    @Test
+    fun shouldRejectIfDirectionAndSwiperProvided () {
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/test/java/io/appium/espressoserver/test/model/MobileSwipeTest.kt]: There is an Expect annotation if a tests expects an exception to be thrown: https://github.com/junit-team/junit4/wiki/exception-testing

File: espresso-server/build.gradle [hunk 10]
```diff
@@ -1,6 +1,7 @@
 // Top-level build file where you can add configuration options common to all sub-projects/modules.
 
 buildscript {
+    ext.kotlin_version = '1.3.11'
```
Comment (mykola-mokhnach) [inline espresso-server/build.gradle]: can we share this variable between both scripts?

File: espresso-server/app/build.gradle [hunk 11]
```diff
@@ -14,8 +16,7 @@ android {
 
     buildTypes {
         release {
-            minifyEnabled false
-            proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
+            minifyEnabled true
```
Comment (mykola-mokhnach) [inline espresso-server/app/build.gradle]: why should we enable that? this setting might corrupt class names, so they will not be visible for reflection calls

**Comment by dpgraham** [comment]
That's right @KazuCocoa ... do you or @mykola-mokhnach have any ideas why that's happening? I'm going to dig into tomorrow to see what happened.

**Comment by mykola-mokhnach** [comment]
@dpgraham Most likely because of Travis slowness. How about moving the tests to Azure CI? @SrinivasanTarget has already done this for java client and the performance is good: 
https://github.com/appium/java-client/pull/1081

File: espresso-server/app/build.gradle [hunk 12]
```diff
@@ -33,28 +35,35 @@ dependencies {
...
+    androidTestImplementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
+    androidTestImplementation "org.jetbrains.kotlin:kotlin-reflect:$kotlin_version"
 }
 
 tasks.withType(Test) {
     systemProperty "skipespressoserver", "true"
-}
\ No newline at end of file
+}
+repositories {
```
Comment (mykola-mokhnach) [inline espresso-server/app/build.gradle]: should we repeat that in both scripts?

**Pull request merged**